mqtt/mqtt-broker/tests/common.rs

```diff
@@ -284,11 +287,135 @@ where
+#[derive(Debug)]
+pub struct PacketStream {
+    codec: Framed<TimeoutStream<TcpStream>, PacketCodec>,
+}
+
+impl PacketStream {
+    /// Creates a client and opens TCP connection to the server.
+    /// No MQTT packets are sent at this moment.
+    pub async fn open(server_addr: impl ToSocketAddrs) -> Self {
+        // broker may not be available immediately in the test,
+        // so we'll try to connect for some time.
+        let mut result = TcpStream::connect(&server_addr).await;
+        let start_time = Instant::now();
+        while let Err(_) = result {
+            if start_time.elapsed() > *DEFAULT_TIMEOUT {
+                break;
+            }
+            result = TcpStream::connect(&server_addr).await;
+        }
+
+        let tcp_stream = result.expect("unable to establish tcp connection");
+        let mut timeout = TimeoutStream::new(tcp_stream);
+        timeout.set_read_timeout(Some(*DEFAULT_TIMEOUT));
+        timeout.set_write_timeout(Some(*DEFAULT_TIMEOUT));
+
+        let codec = Framed::new(timeout, PacketCodec::default());
+
+        Self { codec }
+    }
+
+    /// Creates a client, opens TCP connection to the server,
+    /// and sends CONNECT packet.
+    pub async fn connect(
+        client_id: ClientId,
+        server_addr: impl ToSocketAddrs,
+        username: Option<String>,
+        password: Option<String>,
+    ) -> Self {
+        let mut client = Self::open(server_addr).await;
+        client
+            .send_connect(Connect {
+                username,
+                password,
+                client_id,
+                ..Default::default()
+            })
+            .await;
+        client
+    }
+
+    pub async fn send_connect(&mut self, connect: Connect) {
+        self.send_packet(Packet::Connect(connect)).await;
+    }
+
+    pub async fn send_publish(&mut self, publish: Publish) {
+        self.send_packet(Packet::Publish(publish)).await;
+    }
+
+    pub async fn send_packet(&mut self, packet: Packet) {
+        self.codec
+            .send(packet)
+            .await
+            .expect("Unable to send a packet");
+    }
+}
+
+impl Stream for PacketStream {
+    type Item = Packet;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        match self.codec.poll_next_unpin(cx) {
+            Poll::Ready(Some(result)) => {
+                Poll::Ready(Some(result.expect("Error decoding incoming packet")))
+            }
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// Used to control server lifetime during tests. Implements
+/// Drop to cleanup resources after every test.
+pub struct ServerHandle {
+    address: String,
+    shutdown: Option<Sender<()>>,
+    task: Option<JoinHandle<Result<BrokerState, Error>>>,
+}
+
+impl ServerHandle {
+    pub fn address(&self) -> String {
```

Comment:
```suggestion
    pub fn address(&self) -> &str {
```
seems like it makes unneccessary string allocations..

Reply:
There are some complications with lifetime of &str for spawning broker thread. Will leave this as is.